core/src/commonMain/kotlin/Observers.kt

```diff
@@ -0,0 +1,80 @@
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.MutableSharedFlow
+import kotlinx.coroutines.flow.filter
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.flow.onCompletion
+import kotlinx.coroutines.flow.onSubscription
+import kotlin.coroutines.cancellation.CancellationException
+
+internal expect fun Peripheral.observationHandler(): Observation.Handler
+
+/**
+ * Manages observations for the specified [peripheral].
+ *
+ * The [characteristicChanges] property is expected to be fed with all characteristic changes associated with the
+ * [peripheral]. The changes are then fanned out to individual [Flow]s created via [acquire] (associated with a specific
+ * characteristic).
+ *
+ * For example, if you have a sequence of characteristic changes represented by characteristic A, B and C with their
+ * corresponding change uniquely identified by a change number postfix (in other words: characteristic A emitting 3
+ * different changes would be represented as A1, A2 and A3):
+ *
+ * ```
+ *                                                        .--- acquire(A) --> A1, A2, A3
+ *                             .-----------------------. /
+ *  A1, B1, C1, A2, A3, B2 --> | characteristicChanges | ----- acquire(B) --> B1, B2
+ *                             '-----------------------' \
+ *                                                        '--- acquire(C) --> C1
+ * ```
+ *
+ * @param peripheral to perform notification actions against to enable/disable the observations.
+ */
+internal class Observers<T>(
+    private val peripheral: Peripheral,
+    private val logging: Logging,
+    extraBufferCapacity: Int = 0,
+) {
+
+    val characteristicChanges = MutableSharedFlow<ObservationEvent<T>>(extraBufferCapacity = extraBufferCapacity)
+    private val observations = synchronizedMapOf<Characteristic, Observation>()
+
+    fun acquire(
+        characteristic: Characteristic,
+        onSubscription: OnSubscriptionAction,
+    ): Flow<T> {
+        val handler = peripheral.observationHandler()
+        val observation = observations.synchronized {
+            getOrPut(characteristic) {
+                Observation(peripheral.state, handler, characteristic, logging, peripheral.identifier)
+            }
+        }
+
+        return characteristicChanges
+            .onSubscription { observation.onSubscription(onSubscription) }
+            .filter { event -> event.isAssociatedWith(characteristic) }
+            .map(::dematerialize)
+            .onCompletion { observation.onCompletion(onSubscription) }
+    }
+
+    suspend fun onConnected() {
+        observations.entries.forEach { (characteristic, observation) ->
+            try {
+                observation.onConnected()
+            } catch (cancellation: CancellationException) {
+                throw cancellation
+            } catch (e: Exception) {
+                characteristicChanges.emit(ObservationEvent.Error(characteristic, e))
```

Comment:
It's unclear here what the behavior is regarding the `CancellationException`, if this is as intended can we please have a comment explaining what is going on?

Reply:
Is the following code comment sufficient / what you had in mind?

https://github.com/JuulLabs/kable/blob/b8fd68a8266f384e859cd8bbbdb14f65619b4ee7/core/src/commonMain/kotlin/Observers.kt#L63-L74

Reply:
There is a whole discussion around being able to catch exceptions except for `CancellationException`s: https://github.com/Kotlin/kotlinx.coroutines/issues/1814

Basically, it is unfortunate that we can't use `runCatching` to make this read much more clearly.

Other libraries have introduced their own variants of `runCatching` (e.g. [`runSuspendCatching`](https://github.com/michaelbull/kotlin-result/pull/67)) to make code more readable for this kind of control flow.

Reply:
The bit that is ambiguous (to me at least!) is whether a `CancellationException` is thrown or emitted by this flow.

Reply:
Oh, my understanding is that any suspension point can potentially throw `CancellationException` if the Coroutine it is running in has been cancelled. But to be honest, I'm not 100% confident in that. Like, `ensureActive` function must exist for a reason, so maybe cancellation would have to be explicitly checked for to throw that exception.

Since cancellation is cooperative (and I'm not calling `ensureActive` anywhere) then `CancellationException` may never be thrown here.

I'll look into exactly when `CancellationException` is thrown and follow up.

Reply:
Looking at this again in-situ rather than in the PR diff it makes sense. I think I had a bit of a brain fade there when I first read it but looking at it again now it's fine. I gotta reduce the blood volume in my caffeine system...

Reply:
> I'll look into exactly when `CancellationException` is thrown and follow up.

<details>
<summary>suspend function cancellation (or lack thereof)</summary>

```kotlin
var i = 0

fun main() = runBlocking<Unit> {
    val job = launch(Dispatchers.IO) {
        try {
            while (true) test()
        } catch (cancellation: CancellationException) {
            println("caught cancellation")
            throw cancellation
        }
    }

    delay(5_000L)
    println("cancel")
    job.cancel()
    println("done")
}

private suspend fun test() {
    println(i++)
    Thread.sleep(1_000L)
}
```

Output:

```
0
1
2
3
4
cancel
done
5
6
7
8
9
..
```
</details>

<details>
<summary>withContext cancellation (works)</summary>

```kotlin
var i = 0

fun main() = runBlocking<Unit> {
    val job = launch(Dispatchers.IO) {
        try {
            while (true) test()
        } catch (cancellation: CancellationException) {
            println("caught cancellation")
            throw cancellation
        }
    }

    delay(5_000L)
    println("cancel")
    job.cancel()
    println("done")
}

private suspend fun test() = withContext(Dispatchers.IO) {
    println(i++)
    Thread.sleep(1_000L)
}
```

Output:

```kotlin
0
1
2
3
4
cancel
done
caught cancellation

Process finished with exit code 0
```
</details>

In summary: `suspend` functions on their own do **NOT** automatically check for cancellation, _but_ `withContext` (and some other Coroutines functions, e.g. `ensureActive`) **do** check for cancellation (and throw `CancellationException` if Coroutine context is cancelled).

Deep down in the callstack of `observation.onConnected()`, `withContext` is used, so catching (and re-throwing `CancellationException`) does make sense here.

Reply:
Great article on the subject: https://maxkim.eu/things-every-kotlin-developer-should-know-about-coroutines-part-5-cancellation